scipy/signal/spectral.py

```diff
@@ -511,11 +509,401 @@ def spectrogram(x, fs=1.0, window=('tukey',.25), nperseg=256, noverlap=None,
+        when the employed window function starts at zero. Defaults to True.
+    axis : int, optional
+        Axis along which the STFT is computed; the default is over the last
+        axis (i.e. ``axis=-1``).
+
+    Returns
+    -------
+    f : ndarray
+        Array of sample frequencies.
+    t : ndarray
+        Array of segment times.
+    Zxx : ndarray
+        STFT of x. By default, the last axis of Zxx corresponds to the
+        segment times.
+
+    See Also
+    --------
+    lombscargle: Lomb-Scargle periodogram for unevenly sampled data
+    welch: Power spectral density by Welch's method.
+    spectrogram: Spectrogram by Welch's method.
+    csd: Cross spectral density by Welch's method.
+    istft: Inverse Short Time Fourier Transform
+
+    Notes
+    -----
+    In order to enable inversion of an STFT via the inverse STFT in
+    `istft`, the signal windowing must obey the constraint of "Constant
+    OverLap Add" (COLA), and the input signal must have complete windowing
+    coverage (i.e. (x.shape[axis] - nperseg) % (nperseg-noverlap) == 0).
+
+    The COLA constraint ensures that every point in the input data is
+    equally weighted, thereby avoiding aliasing and allowing full
+    reconstruction. Whether a choice of `window`, `nperseg`, and `noverlap`
+    satisfy this constraing can be tested with
+    `scipy.signal.spectral.check_COLA`.
+
+    ..  versionadded:: 0.18.0
+
+    References
+    ----------
+    .. [1] Oppenheim, Alan V., Ronald W. Schafer, John R. Buck
+           "Discrete-Time Signal Processing", Prentice Hall, 1999.
+    .. [2] Daniel W. Griffin, Jae S. Limdt "Signal Estimation from Modified
+           Short Fourier Transform", IEEE 1984, 10.1109/TASSP.1984.1164317
+
+    Examples
+    --------
+    >>> from scipy import signal
+    >>> from scipy.fftpack import fftshift
+    >>> import matplotlib.pyplot as plt
+
+    Generate a test signal, a 2 Vrms sine wave whose frequency linearly
+    changes with time from 1kHz to 2kHz, corrupted by 0.001 V**2/Hz of
+    white noise sampled at 10 kHz.
+
+    >>> fs = 10e3
+    >>> N = 1e5
+    >>> amp = 2 * np.sqrt(2)
+    >>> noise_power = 0.001 * fs / 2
+    >>> time = np.arange(N) / fs
+    >>> freq = np.linspace(1e3, 2e3, N)
+    >>> x = amp * np.sin(2*np.pi*freq*time)
+    >>> x += np.random.normal(scale=np.sqrt(noise_power), size=time.shape)
+
+    Compute and plot the STFT's magnitude and phase
+
+    >>> f, t, Zxx = signal.stft(x, fs, nperseg=1000)
+    >>> f = fftshift(f)
+    >>> Zxx = fftshift(Zxx, axes=0)
+    >>> fig, ax = plt.subplots(nrows=2, sharex=True)
+    >>> ax[0].pcolormesh(t, f, np.abs(Zxx))
+    >>> ax[1].pcolormesh(t, f, np.angle(Zxx))
+    >>> ax[0].set_title('STFT Magnitude')
+    >>> ax[0].set_ylabel('Frequency [Hz]')
+    >>> ax[0].set_title('STFT Phase')
+    >>> ax[1].set_ylabel('Frequency [Hz]')
+    >>> ax[1].set_xlabel('Time [sec]')
+    >>> plt.show()
+    """
```

Comment:
Sorry if this is out of place, but I've been trying your example, and in the magnitude plot I get a maximum frequency of 3kHz, which seams like it is different from the 2kHz max on the input. Is this the expected output (i.e. have I misunderstood the code).
I'm trying to verify this with the example in istft, but there the sample frequency lower than the max input, meaning that the returned Zxx is "wrap"/aliased.